File: scripts/catalog.py
#!/usr/bin/env python3
"""z-yap catalog — structured summary of all zettels with their connections.

Usage:
  python3 scripts/catalog.py             # full catalog
  python3 scripts/catalog.py --compact   # one-line-per-zettel
  python3 scripts/catalog.py --markdown  # plain markdown (for dist/)
  python3 scripts/catalog.py types       # filter by keyword
"""

import signal
import sys
from pathlib import Path

signal.signal(signal.SIGPIPE, signal.SIG_DFL)

ROOT = Path(__file__).parent.parent
ZETTELS_DIR = ROOT / "zettels"
CONNECTIONS_FILE = ROOT / "connections.cypher"

sys.path.insert(0, str(Path(__file__).parent))
from lib.zettel import load_all, Zettel
from lib.connections import parse_file, edges_for


def main() -> None:
    args = sys.argv[1:]
    compact = "--compact" in args
    markdown = "--markdown" in args
    filters = [a for a in args if not a.startswith("--")]

    zettels = load_all(ZETTELS_DIR)
    edges = parse_file(CONNECTIONS_FILE) if CONNECTIONS_FILE.exists() else []

    if filters:
        zettels = [z for z in zettels if _matches(z, filters)]

    if markdown:
        _print_markdown(zettels, edges, compact)
    elif compact:
        _print_compact(zettels)
    else:
        _print_full(zettels, edges)


def _matches(z: Zettel, filters: list[str]) -> bool:
    searchable = f"{z.title} {z.slug} {' '.join(z.tags)}".lower()
    return any(f.lower() in searchable for f in filters)


def _print_markdown(zettels: list[Zettel], edges, compact: bool) -> None:
    print(f"# z-yap Catalog ({len(zettels)} zettels)\n")
    for z in zettels:
        if compact:
            tags = ", ".join(z.tags)
            print(f"- **{z.title}** `{z.slug}` [{tags}]")
        else:
            print(f"## {z.title}")
            print(f"`{z.slug}` — tags: {', '.join(z.tags)}")
            out, inc = edges_for(z.slug, edges)
            if out:
                print(f"\n**Outgoing:**")
                for e in out:
                    note = f" — {e.note}" if e.note else ""
                    print(f"- [{e.label}] → [[{e.target}]]{note}")
            if inc:
                print(f"\n**Incoming:**")
                for e in inc:
                    note = f" — {e.note}" if e.note else ""
                    print(f"- [[{e.source}]] [{e.label}] →{note}")
            print()


def _print_compact(zettels: list[Zettel]) -> None:
    max_title = max((len(z.title) for z in zettels), default=0)
    max_slug = max((len(z.slug) for z in zettels), default=0)
    print(f"{'Title':<{max_title}}  {'Slug':<{max_slug}}  Tags")
    print(f"{'─' * max_title}  {'─' * max_slug}  ────")
    for z in zettels:
        tags = ", ".join(z.tags)
        print(f"{z.title:<{max_title}}  {z.slug:<{max_slug}}  {tags}")


def _print_full(zettels: list[Zettel], edges) -> None:
    print(f"# z-yap Catalog ({len(zettels)} zettels)\n")
    for z in zettels:
        print(f"  {z.title}")
        print(f"  slug: {z.slug}")
        print(f"  tags: {', '.join(z.tags)}")
        out, inc = edges_for(z.slug, edges)
        if out:
            for e in out:
                note = f" — {e.note}" if e.note else ""
                print(f"    → [{e.label}] [[{e.target}]]{note}")
        if inc:
            for e in inc:
                note = f" — {e.note}" if e.note else ""
                print(f"    ← [[{e.source}]] [{e.label}]{note}")
        print()


if __name__ == "__main__":
    main()

File: scripts/catalog.js
#!/usr/bin/env node
/**
 * Zettel inventory with optional filtering.
 *
 * Usage:
 *   node scripts/catalog.js                   # full catalog (ANSI)
 *   node scripts/catalog.js --tag solver      # filter by tag
 *   node scripts/catalog.js --status planned  # filter by epistemic status
 *   node scripts/catalog.js --search fuzzy    # text search on slug + title
 *   node scripts/catalog.js --markdown        # plain markdown (for dist/)
 *   node scripts/catalog.js --compact         # one-line-per-zettel (ANSI)
 */

import { loadAll, parseConnections, edgesFor } from "./lib/parse.js";
import { c } from "./lib/colors.js";

const EPISTEMIC = new Set([
  "implemented", "in-progress", "planned", "speculative",
  "deprecated", "rejected", "deferred", "incomplete",
]);
const READINESS = new Set(["ready", "blocked", "needs-design"]);
const STRUCTURAL = new Set(["thread", "queue", "hub", "adr"]);

const args = process.argv.slice(2);
const markdown = args.includes("--markdown");
const compact = args.includes("--compact") || markdown;
const filters = {};
for (let i = 0; i < args.length; i++) {
  if (args[i].startsWith("--") && args[i + 1] && !args[i + 1].startsWith("--")) {
    filters[args[i].slice(2)] = args[++i];
  }
}

let zettels = loadAll();
if (filters.tag)    zettels = zettels.filter((z) => z.tags.includes(filters.tag));
if (filters.status) zettels = zettels.filter((z) => z.tags.includes(filters.status));
if (filters.search) {
  const q = filters.search.toLowerCase();
  zettels = zettels.filter((z) => z.slug.toLowerCase().includes(q) || z.title.toLowerCase().includes(q));
}

const pad = (s, w) => (s.length >= w ? s : s + " ".repeat(w - s.length));

function statusColor(s) {
  if (s === "implemented") return c.green;
  if (["deprecated", "rejected"].includes(s)) return c.red;
  if (["blocked", "needs-design"].includes(s)) return c.yellow;
  if (["in-progress", "incomplete"].includes(s)) return c.cyan;
  return c.reset;
}

if (markdown) {
  printMarkdown(zettels);
} else if (compact) {
  printCompact(zettels);
} else {
  printFull(zettels);
}

function printMarkdown(zettels) {
  console.log(`# z-yap Catalog (${zettels.length} zettels)\n`);
  for (const z of zettels) {
    const tags = z.tags.join(", ");
    console.log(`- **${z.title}** \`${z.slug}\` [${tags}]`);
  }
}

function printCompact(zettels) {
  const slugW = Math.max(4, ...zettels.map((z) => z.slug.length));
  const titleW = 40;
  console.log(`\n${c.bold}CATALOG${c.reset} — ${zettels.length} zettels`);
  if (Object.keys(filters).length) {
    console.log(c.gray + Object.entries(filters).map(([k, v]) => `${k}=${v}`).join("  ") + c.reset);
  }
  console.log("─".repeat(slugW + titleW + 30));

  for (const z of zettels) {
    const epistemic = z.tags.find((t) => EPISTEMIC.has(t)) || "—";
    console.log(
      `${c.cyan}${pad(z.slug, slugW)}${c.reset}  ` +
      `${pad(z.title.slice(0, titleW), titleW)}  ` +
      `${statusColor(epistemic)}${epistemic}${c.reset}`
    );
  }
  console.log("");
}

function printFull(zettels) {
  const edges = parseConnections();
  const slugW = Math.max(4, ...zettels.map((z) => z.slug.length));
  const titleW = 40;

  console.log(`\n${c.bold}CATALOG${c.reset} — ${zettels.length} zettels`);
  if (Object.keys(filters).length) {
    console.log(c.gray + Object.entries(filters).map(([k, v]) => `${k}=${v}`).join("  ") + c.reset);
  }
  console.log("─".repeat(slugW + titleW + 30));

  for (const z of zettels) {
    const structural = z.tags.find((t) => STRUCTURAL.has(t)) || "";
    const epistemic = z.tags.find((t) => EPISTEMIC.has(t)) || "—";
    const readiness = z.tags.find((t) => READINESS.has(t));
    const stStr = readiness ? `${epistemic} [${readiness}]` : epistemic;

    console.log(
      `${c.cyan}${pad(z.slug, slugW)}${c.reset}  ` +
      `${pad(z.title.slice(0, titleW), titleW)}  ` +
      `${structural ? c.magenta + pad(structural, 8) + c.reset + "  " : "          "}` +
      `${statusColor(epistemic)}${stStr}${c.reset}`
    );
  }
  console.log("");
}
